include/irr/builtin/glsl/bxdf/brdf/specular/beckmann_smith.glsl

```diff
@@ -96,20 +102,34 @@ irr_glsl_BSDFSample irr_glsl_beckmann_smith_cos_gen_sample(in irr_glsl_Anisotrop
     float HdotV = dot(H,interaction.isotropic.V.dir);
     irr_glsl_BSDFSample smpl;
     smpl.L = H*2.0*HdotV - interaction.isotropic.V.dir;
-
-    //==== compute probability ====
-    //PBRT does it like a2 = cos2phi*ax*ax + sin2phi*ay*ay
-    float a2 = ax*ay;
-    float lambda = irr_glsl_smith_beckmann_Lambda(irr_glsl_smith_beckmann_C2(interaction.isotropic.NdotV_squared, a2));
-    float G1 = 1.0 / (1.0 + lambda);
-    smpl.probability = irr_glsl_beckmann(a2,NdotH*NdotH) * G1 * abs(dot(interaction.isotropic.V.dir,H)) / interaction.isotropic.NdotV;
+	smpl.NdotH = NdotH;
+	smpl.VdotH = HdotV;
+	//TODO smpl.Ldot[T|B|N]
 
     return smpl;
 }
-irr_glsl_BSDFSample irr_glsl_beckmann_smith_cos_gen_sample(in irr_glsl_AnisotropicViewSurfaceInteraction interaction, in uvec2 _sample, in float _ax, in float _ay)
+irr_glsl_BSDFSample irr_glsl_beckmann_smith_cos_generate(in irr_glsl_AnisotropicViewSurfaceInteraction interaction, in uvec2 _sample, in float _ax, in float _ay)
 {
     vec2 u = vec2(_sample)/float(UINT_MAX);
-    return irr_glsl_beckmann_smith_cos_gen_sample(interaction, u, _ax, _ay);
+    return irr_glsl_beckmann_smith_cos_generate(interaction, u, _ax, _ay);
+}
+
+//TODO remove interaction param when irr_glsl_BSDFAnisotropicParams contains irr_glsl_AnisotropicViewSurfaceInteraction
+vec3 irr_glsl_beckmann_smith_cos_remainder_and_pdf(out float pdf, in irr_glsl_BSDFAnisotropicParams params, in irr_glsl_AnisotropicViewSurfaceInteraction interaction, in uvec2 u, in mat2x3 ior2, in float ax, in float ay)
+{
+	irr_glsl_BSDFSample s = irr_glsl_beckmann_smith_cos_generate(interaction, u, ax, ay);
+	
+	float a2 = ax*ay;
+	float NdotL2 = s.LdotN*s.LdotN;
+	float lambda_V = irr_glsl_smith_beckmann_Lambda(irr_glsl_smith_beckmann_C2(interaction.isotropic.NdotV_squared, a2));
+	float lambda_L = irr_glsl_smith_beckmann_Lambda(irr_glsl_smith_beckmann_C2(NdotL2, a2));
+	float onePlusLambda_V = 1.0 + lambda_V;
+	float G = 1.0 / onePlusLambda_V;
+	pdf = irr_glsl_beckmann(a2,s.NdotH*s.NdotH)*G*abs(dot(interaction.isotropic.V.dir,H))/interaction.isotropic.NdotV;
+	G = onePlusLambda_V/(onePlusLambda_V+lambda_L);//remainder
+	
+	vec3 fr = irr_glsl_fresnel_conductor(ior2[0], ior2[1], s.VdotH);
+	return fr*G / (4.0 * params.isotropic.interaction.NdotV);
```

Comment:
I'm gonna need an explainer on this...

btw are you sure that the Sampling of visible distribution does not remove the `1.0/(4.0*NdotV)` factor from the remainder like the GGX does?